### PR TITLE
[GNRC PPP] PPP header definition and helpers

### PR DESCRIPTION
**Pull request by jia200x**
Hi.

I'm writing a PPP stack for RIOT (for usage with mobile communications, etc).

This is a PR of PPP header definition and running tests.
Best regards

### CODE REVIEW

**Comment by authmillenon** [comment]
Please inform me, when this PR isn't WIP anymore.

File: sys/include/net/ppptype.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#endif
+
+
+#define PPPTYPE_IPV6       (0x0057)    /**< IPv6 packet in PPP*/
+#define PPPTYPE_IPV4       (0x0021) /**< IPv4 packet in PPP */
+#define PPPTYPE_NCP_IPV6       (0x8057)    /**< NCP packet for IPV6*/
+#define PPPTYPE_NCP_IPV4       (0x8021)    /**< NCP packet for IPv4*/
+#define PPPTYPE_LCP       (0xC021)    /**< LCP packet  */
+#define PPPTYPE_PAP       (0xC023)    /**< PAP packet  */
+#define PPPTYPE_UNKNOWN       (0x8001)    /**<Unknown  packet (Unused protocol number according to IANA) */
```
Comment (miri64) [inline sys/include/net/ppptype.h]: Alignment is all over the place in all of these lines.

**Comment by miri64** [comment]
Tests run stand-alone and with the other tests, apart from the indentation issue I'm fine with the code. @OlegHahm what was the reason for the WIP label?

**Comment by kYc0o** [comment]
this PR still marked as WIP, so I think is better to give a good review for the next release.

**Comment by OlegHahm** [comment]
> @OlegHahm what was the reason for the WIP label?

The title. ;)

File: sys/include/net/ppp/hdr.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2016 José Ignacio Alamos <jialamos@uc.cl>
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: Are you sure this doesn't need to be "INRIA"?

File: sys/include/net/ppp/hdr.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ *
+ *  0                   1                   2                   3
+ *  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |     Code      |  Identifier   |            Length             |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |    Payload ...
+ * +-+-+-+-+
+ *
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: Can you maybe put a reference to the RFC or something here? See [`ipv6_hdr_t`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ipv6.h#L38-L40) for example.

File: sys/include/net/ppp/hdr.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Get code field of PPP packet
+ *
+ * @param[in] hdr    PPP header
+ *
+ * @return    code of PPP header
+ */
+static inline uint8_t ppp_hdr_get_code(ppp_hdr_t *hdr)
+{
+    return hdr->code;
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: This function seems awfully unnecessary.

File: sys/include/net/ppp/hdr.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Set code field of PPP header
+ *
+ * @param[in] hdr    PPP header
+ * @param[in] code    code to be set
+ *
+ */
+static inline void ppp_hdr_set_code(ppp_hdr_t *hdr, uint8_t code)
+{
+    hdr->code = code;
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: dito

File: sys/include/net/ppp/hdr.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Get id field of PPP header
+ *
+ * @param[in] hdr    PPP header
+ *
+ * @return    id of PPP header
+ */
+static inline uint8_t ppp_hdr_get_id(ppp_hdr_t *hdr)
+{
+    return hdr->id;
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: dito

File: sys/include/net/ppp/hdr.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Set id field of PPP header
+ *
+ * @param[in] hdr    PPP header
+ * @param[in] id    id to be set
+ *
+ */
+static inline void ppp_hdr_set_id(ppp_hdr_t *hdr, uint8_t id)
+{
+    hdr->id = id;
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: dito

File: sys/include/net/ppp/hdr.h [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Get length field of PPP header
+ *
+ * @param[in] hdr    PPP header
+ *
+ * @return    length of PPP header
+ */
+static inline uint16_t ppp_hdr_get_length(ppp_hdr_t *hdr)
+{
+    return byteorder_ntohs(hdr->length);
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: dito

File: sys/include/net/ppp/hdr.h [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief    Set length field of PPP header
+ *
+ * @param[in] hdr    PPP header
+ * @param[in] length    length to be set
+ *
+ */
+static inline void ppp_hdr_set_length(ppp_hdr_t *hdr, uint16_t length)
+{
+    hdr->length = byteorder_htons(length);
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: dito

File: sys/include/net/ppptype.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @defgroup    net_ppptypes PPP types
+ * @ingroup     net
+ * @brief       PPP types
+
```
Comment (miri64) [inline sys/include/net/ppptype.h]: Link to the database might be helpful for future extensions. See [ethertype.h](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ethertype.h#L13-L15) for reference

**Comment by miri64** [comment]
@jia200x ping?

File: sys/include/net/ppptype.h (hunk 10 above)
Comment (miri64) [inline sys/include/net/ppptype.h]: :ballot_box_with_check:

File: sys/include/net/ppp/hdr.h (hunk 3 above)
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: :ballot_box_with_check:

File: sys/include/net/ppp/hdr.h (hunk 2 above)
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: Ping

**Comment by jia200x** [comment]
I think I have addressed all issues (I'm waiting for the answer of the proper copyright)

**Comment by miri64** [comment]
When you have one, please squash!

File: sys/include/net/ppp/hdr.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+ * +-+-+-+-+
+ *
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc1661#section-5">
+ *          RFC 1661, section 5
+ *      </a>
+ */
+/*  PPP pkt header struct */
+typedef struct __attribute__((packed)){
+    uint8_t code;
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]: Please document the fields individually (to pass murdock).

**Comment by miri64** [comment]
Something went wrong on rebase oO. Sorry I can't merge it like this. Postponed.

**Comment by jia200x** [comment]
@miri64 I have been trying to solve this for a while. Any clue of how to proceed?
Rebasing from master sends several conflicts.

**Comment by miri64** [comment]
That can't be! The very idea of a rebase to master is to *resolve* conflicts. Are you sure you have the most current master?

```bash
git checkout master
git pull
git checkout PPP_hdr
git rebase -i master   # in the file opened "Added PPP header definitions" should be contained in the first line
git push -f
```

**Comment by jia200x** [comment]
@miri64 yes, it's the latest master. For some reason is giving conflicts.
I'm trying to deal with it.

**Comment by jia200x** [comment]
@miri64 close enough... (reverted to the original state). I will try to rebase again

**Comment by miri64** [comment]
:+1:

**Comment by jia200x** [comment]
@miri64 finally solved.
Cheers

**Comment by kYc0o** [comment]
@jia200x Please fix the doxygen errors Murdock found. Besides it looks ok but we really need to find a solution for this DTLS issue :/

File: sys/include/net/ppp/hdr.h [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    uint8_t code;               /**< Code of PPP packet*/
+    uint8_t id;                 /**< Identifier PPP of packet*/
+    network_uint16_t length;    /**< Length of PPP packet including payload*/
+} ppp_hdr_t;
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* PPP_HDR_H_ */
```
Comment (miri64) [inline sys/include/net/ppp/hdr.h]:
```C
/** @} */
```

missing (that's what Murdock is complaining about ;-))

**Comment by jia200x** [comment]
Hi @kYc0o,

Which DTLS issue do you mean?
Cheers!

**Comment by kYc0o** [comment]
This error doesn't come from your commits, it's about random link errors when building the tinyDTLS package, but that's a separated issue (#6036).

**Comment by jia200x** [comment]
@miri64 addressed the issue.
Cheers!

**Comment by kYc0o** [comment]
Great! the only error is this DTLS linking, so I can say it's ready. @miri64 can you give your ACK? ;-)

**Comment by kYc0o** [comment]
Oh! just realised that this is not going into this release...

**Comment by jia200x** [comment]
it was originally going into this release but I had a problem with squashing. @miri64 can we move it again to this release? :-)

**Comment by miri64** [comment]
The release is already feature freezed, so no (I also don't see much sense in merging the headers without any actual implementation attached via backport). But we can merge it into master as soon as murdock is hapy ;-)

**Comment by miri64** [comment]
Ah, that was just the tinyDTLS issue... working on it.

**Pull request merged**